x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/ingest/GeoTileProcessor.java

```diff
@@ -0,0 +1,421 @@
+            return zoom + "/" + x + "/" + y;
+        }
+
+        @Override
+        String getParent() {
+            return zoom == 0 ? "" : hash(zoom - 1, x / 2, y / 2);
+        }
+
+        @Override
+        int getDepth() {
+            return zoom;
+        }
+    }
+
+    static class GeohashHandler extends TileHandler {
+        private final String hash;
+
+        GeohashHandler(String hash) {
+            this.hash = hash;
+        }
+
+        @Override
+        Geometry makePolygon() {
+            return makePolygonFrom(Geohash.toBoundingBox(hash));
+        }
+
+        @Override
+        String getParent() {
+            return hash.length() == 0 ? "" : hash.substring(0, hash.length() - 1);
+        }
+
+        @Override
+        String[] makeChildren() {
+            return Geohash.getSubGeohashes(hash);
+        }
+
+        @Override
+        int getDepth() {
+            return hash.length();
+        }
+    }
+
+    static class GeohexHandler extends TileHandler {
+        private final long h3;
+
+        GeohexHandler(String spec) {
+            this.h3 = parseH3Specification(spec);
+        }
+
+        @Override
+        Geometry makePolygon() {
+            return makePolygonFromH3(h3);
+        }
+
+        @Override
+        String getParent() {
+            int res = H3.getResolution(h3);
+            if (res == 0) {
+                return "";
+            }
+            long parent = H3.h3ToParent(h3);
+            return H3.h3ToString(parent);
+        }
+
+        @Override
+        String[] makeChildren() {
+            return asStringArray(H3.h3ToChildren(h3));
+        }
+
+        @Override
+        String[] makeNonChildren() {
+            return asStringArray(H3.h3ToNoChildrenIntersecting(h3));
+        }
+
+        @Override
+        int getDepth() {
+            return H3.getResolution(h3);
+        }
+
+        private static Geometry makePolygonFromH3(long h3) {
+            return H3CartesianUtil.getNormalizeGeometry(h3);
+        }
+
+        private static String[] asStringArray(long[] cells) {
+            String[] addresses = new String[cells.length];
+            for (int i = 0; i < cells.length; i++) {
+                addresses[i] = H3.h3ToString(cells[i]);
+            }
+            return addresses;
+        }
+
+        private static long parseH3Specification(String spec) {
+            if (spec.toLowerCase(Locale.ROOT).trim().startsWith("h3.")) {
+                return h3FromH3Function(spec.substring(3).split("[,()]"));
+            } else {
+                return H3.stringToH3(spec);
+            }
+        }
+
+        private static long h3FromH3Function(String[] parts) {
+            String function = parts[0].toLowerCase(Locale.ROOT).trim();
+            return switch (function) {
+                // H3.geoToH3(lat,lng,res)
+                case "geotoh3" -> {
+                    assertParams(function, 3, parts.length - 1);
```

Comment:
I think this is a bit hacky and not sure if we should support it this way. This seems to me that should be done at painless level.

Reply:
Done